pkg/crypto/certloader/watcher.go

```diff
@@ -0,0 +1,239 @@
+	if w.FileReloader.certFile != "" {
+		keypairMap[w.FileReloader.certFile] = struct{}{}
+	}
+	if w.FileReloader.privkeyFile != "" {
+		keypairMap[w.FileReloader.privkeyFile] = struct{}{}
+	}
+	for _, path := range w.FileReloader.caFiles {
+		caMap[path] = struct{}{}
+	}
+
+	// prepare the ready channel to be returned. We will close it exactly once.
+	var once sync.Once
+	ready := make(chan struct{})
+	go func() {
+		defer w.fswatcher.Close()
+		for {
+			select {
+			case event := <-w.fswatcher.Events:
+				path := event.Name
+				log := w.log.WithFields(logrus.Fields{
+					logfields.Path: path,
+					"operation":    event.Op,
+				})
+				log.Debug("Received fsnotify event")
+				_, keypairChanged := keypairMap[path]
+				_, caChanged := caMap[path]
+				createOperation := event.Op&fsnotify.Create == fsnotify.Create
+				if (keypairChanged || caChanged) && createOperation {
+					// A file we are interested in has been created. We might
+					// have been watching only over its directory until now,
+					// add it to the watcher since it exists.
+					if err := w.fswatcher.Add(path); err != nil {
+						log.WithError(err).Warnf("Failed to add %s to fsnotify watcher", logfields.Path)
+					}
+				}
+				if keypairChanged {
+					keypair, err := w.ReloadKeypair()
+					if err != nil {
+						log.WithError(err).Warn("keypair update failed")
+						continue
+					}
+					id := keypairId(keypair)
+					log.WithField("keypair-id", id).Info("keypair updated")
+				} else if caChanged {
+					if _, err := w.ReloadCertificateAuthority(); err != nil {
+						log.WithError(err).Warn("certificate authority update failed")
+						continue
+					}
+					log.Info("certificate authority updated")
+				} else {
+					log.Debug("unknown file, ignoring.")
+					continue
+				}
+				if w.Ready() {
+					once.Do(func() {
+						close(ready)
+					})
+				}
+			case err := <-w.fswatcher.Errors:
+				w.log.WithError(err).Warn("fsnotify watcher error")
+			case <-w.stop:
+				w.log.Info("Stopping fsnotify watcher")
+				return
+			}
+		}
+	}()
+
+	return ready
+}
+
+// Stop watching the files.
+func (w *Watcher) Stop() {
+	close(w.stop)
+}
+
+// newFsWatcher returns a fsnotify.Watcher watching over the given files.
+// When a file doesn't exists, we watch over its directory instead to pick up
+// the file creation event.
+func newFsWatcher(caFiles []string, certFile, privkeyFile string) (*fsnotify.Watcher, error) {
+	fswatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	add := func(path string) error {
+		if path == "" {
+			return nil
+		}
+		if err := fswatcher.Add(path); err != nil {
+			// the file might not exists yet, try watching its directory
+			// instead.
+			dir := filepath.Dir(path)
+			if err := fswatcher.Add(dir); err != nil {
+				// complain about the directory, as this will be more useful than
+				// the file itself.
+				return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", dir, err)
```

Comment:
Please, don't capitalize error strings (see https://github.com/golang/go/wiki/CodeReviewComments#error-strings).
Also, do we want to embed the error with `%w` here maybe?